// ng-wormhole/fesm5/ng-wormhole.js
import { Directive, Input, ViewContainerRef, TemplateRef, Inject, NgModule } from '@angular/core';
import { DOCUMENT } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @template T
 * @param {?} target
 * @param {?} node
 * @return {?}
 */
function appendReducer(target, node) {
    target.appendChild(node);
    return target;
}
var NgWormholeDirective = /** @class */ (function () {
    function NgWormholeDirective(viewContainerRef, templateRef, document) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
        this.ngWormholeRenderInPlace = false;
        this.initialized = false;
        this.document = document;
    }
    Object.defineProperty(NgWormholeDirective.prototype, "destinationElement", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.ngWormholeRenderInPlace) {
                return (this.viewContainerRef.element &&
                    this.viewContainerRef.element.nativeElement);
            }
            if (!this.ngWormhole) {
                return null;
            }
            return typeof this.ngWormhole === 'string'
                ? this.document.querySelector(this.ngWormhole)
                : this.ngWormhole;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgWormholeDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.embeddedViewRef = this.viewContainerRef.createEmbeddedView(this.templateRef);
    };
    /**
     * @return {?}
     */
    NgWormholeDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.appendToDestination();
        this.initialized = true;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgWormholeDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.initialized) {
            this.appendToDestination();
        }
    };
    /**
     * @return {?}
     */
    NgWormholeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.embeddedViewRef != null) {
            this.embeddedViewRef.destroy();
        }
    };
    /**
     * @return {?}
     */
    NgWormholeDirective.prototype.appendToDestination = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var startingActiveElement = this.getActiveElement();
        /** @type {?} */
        var destinationElement = this.destinationElement;
        if (destinationElement == null) {
            return;
        }
        /** @type {?} */
        var content = this.embeddedViewRef.rootNodes.reduce(appendReducer, this.document.createDocumentFragment());
        if (destinationElement.nodeType === Node.COMMENT_NODE) {
            destinationElement.parentNode.insertBefore(content, destinationElement.nextSibling);
        }
        else {
            destinationElement.appendChild(content);
        }
        /** @type {?} */
        var resultingActiveElement = this.getActiveElement();
        if (startingActiveElement != null &&
            resultingActiveElement !== startingActiveElement) {
            (/** @type {?} */ (startingActiveElement)).focus();
        }
    };
    /**
     * @return {?}
     */
    NgWormholeDirective.prototype.getActiveElement = /**
     * @return {?}
     */
    function () {
        return this.document.activeElement;
    };
    NgWormholeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngWormhole]' // tslint:disable-line
                },] },
    ];
    /** @nocollapse */
    NgWormholeDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: TemplateRef },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    NgWormholeDirective.propDecorators = {
        ngWormholeRenderInPlace: [{ type: Input }],
        ngWormhole: [{ type: Input }]
    };
    return NgWormholeDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgWormholeModule = /** @class */ (function () {
    function NgWormholeModule() {
    }
    NgWormholeModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [NgWormholeDirective],
                    exports: [NgWormholeDirective]
                },] },
    ];
    return NgWormholeModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgWormholeDirective, NgWormholeModule };
