// ng-wormhole/bundles/ng-wormhole.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-wormhole', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['ng-wormhole'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /**
     * @template T
     * @param {?} target
     * @param {?} node
     * @return {?}
     */
    function appendReducer(target, node) {
        target.appendChild(node);
        return target;
    }
    var NgWormholeDirective = (function () {
        function NgWormholeDirective(viewContainerRef, templateRef, document) {
            this.viewContainerRef = viewContainerRef;
            this.templateRef = templateRef;
            this.ngWormholeRenderInPlace = false;
            this.initialized = false;
            this.document = document;
        }
        Object.defineProperty(NgWormholeDirective.prototype, "destinationElement", {
            get: /**
             * @return {?}
             */ function () {
                if (this.ngWormholeRenderInPlace) {
                    return (this.viewContainerRef.element &&
                        this.viewContainerRef.element.nativeElement);
                }
                if (!this.ngWormhole) {
                    return null;
                }
                return typeof this.ngWormhole === 'string'
                    ? this.document.querySelector(this.ngWormhole)
                    : this.ngWormhole;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NgWormholeDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.embeddedViewRef = this.viewContainerRef.createEmbeddedView(this.templateRef);
            };
        /**
         * @return {?}
         */
        NgWormholeDirective.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                this.appendToDestination();
                this.initialized = true;
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        NgWormholeDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (this.initialized) {
                    this.appendToDestination();
                }
            };
        /**
         * @return {?}
         */
        NgWormholeDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                if (this.embeddedViewRef != null) {
                    this.embeddedViewRef.destroy();
                }
            };
        /**
         * @return {?}
         */
        NgWormholeDirective.prototype.appendToDestination = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var startingActiveElement = this.getActiveElement();
                /** @type {?} */
                var destinationElement = this.destinationElement;
                if (destinationElement == null) {
                    return;
                }
                /** @type {?} */
                var content = this.embeddedViewRef.rootNodes.reduce(appendReducer, this.document.createDocumentFragment());
                if (destinationElement.nodeType === Node.COMMENT_NODE) {
                    destinationElement.parentNode.insertBefore(content, destinationElement.nextSibling);
                }
                else {
                    destinationElement.appendChild(content);
                }
                /** @type {?} */
                var resultingActiveElement = this.getActiveElement();
                if (startingActiveElement != null &&
                    resultingActiveElement !== startingActiveElement) {
                    ((startingActiveElement)).focus();
                }
            };
        /**
         * @return {?}
         */
        NgWormholeDirective.prototype.getActiveElement = /**
         * @return {?}
         */
            function () {
                return this.document.activeElement;
            };
        NgWormholeDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[ngWormhole]' // tslint:disable-line
                    },] },
        ];
        /** @nocollapse */
        NgWormholeDirective.ctorParameters = function () {
            return [
                { type: core.ViewContainerRef },
                { type: core.TemplateRef },
                { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
            ];
        };
        NgWormholeDirective.propDecorators = {
            ngWormholeRenderInPlace: [{ type: core.Input }],
            ngWormhole: [{ type: core.Input }]
        };
        return NgWormholeDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgWormholeModule = (function () {
        function NgWormholeModule() {
        }
        NgWormholeModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [NgWormholeDirective],
                        exports: [NgWormholeDirective]
                    },] },
        ];
        return NgWormholeModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgWormholeDirective = NgWormholeDirective;
    exports.NgWormholeModule = NgWormholeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
