// ng-wormhole/esm2015/lib/ng-wormhole.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, TemplateRef, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
/**
 * @template T
 * @param {?} target
 * @param {?} node
 * @return {?}
 */
function appendReducer(target, node) {
    target.appendChild(node);
    return target;
}
export class NgWormholeDirective {
    /**
     * @param {?} viewContainerRef
     * @param {?} templateRef
     * @param {?} document
     */
    constructor(viewContainerRef, templateRef, document) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
        this.ngWormholeRenderInPlace = false;
        this.initialized = false;
        this.document = document;
    }
    /**
     * @return {?}
     */
    get destinationElement() {
        if (this.ngWormholeRenderInPlace) {
            return (this.viewContainerRef.element &&
                this.viewContainerRef.element.nativeElement);
        }
        if (!this.ngWormhole) {
            return null;
        }
        return typeof this.ngWormhole === 'string'
            ? this.document.querySelector(this.ngWormhole)
            : this.ngWormhole;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.embeddedViewRef = this.viewContainerRef.createEmbeddedView(this.templateRef);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.appendToDestination();
        this.initialized = true;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.initialized) {
            this.appendToDestination();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.embeddedViewRef != null) {
            this.embeddedViewRef.destroy();
        }
    }
    /**
     * @return {?}
     */
    appendToDestination() {
        /** @type {?} */
        let startingActiveElement = this.getActiveElement();
        /** @type {?} */
        let destinationElement = this.destinationElement;
        if (destinationElement == null) {
            return;
        }
        /** @type {?} */
        let content = this.embeddedViewRef.rootNodes.reduce(appendReducer, this.document.createDocumentFragment());
        if (destinationElement.nodeType === Node.COMMENT_NODE) {
            destinationElement.parentNode.insertBefore(content, destinationElement.nextSibling);
        }
        else {
            destinationElement.appendChild(content);
        }
        /** @type {?} */
        let resultingActiveElement = this.getActiveElement();
        if (startingActiveElement != null &&
            resultingActiveElement !== startingActiveElement) {
            (/** @type {?} */ (startingActiveElement)).focus();
        }
    }
    /**
     * @return {?}
     */
    getActiveElement() {
        return this.document.activeElement;
    }
}
NgWormholeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngWormhole]' // tslint:disable-line
            },] },
];
/** @nocollapse */
NgWormholeDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
NgWormholeDirective.propDecorators = {
    ngWormholeRenderInPlace: [{ type: Input }],
    ngWormhole: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgWormholeDirective.prototype.ngWormholeRenderInPlace;
    /** @type {?} */
    NgWormholeDirective.prototype.ngWormhole;
    /** @type {?} */
    NgWormholeDirective.prototype.embeddedViewRef;
    /** @type {?} */
    NgWormholeDirective.prototype.initialized;
    /** @type {?} */
    NgWormholeDirective.prototype.document;
    /** @type {?} */
    NgWormholeDirective.prototype.viewContainerRef;
    /** @type {?} */
    NgWormholeDirective.prototype.templateRef;
}
